Redesign NATS module to support both servers (#16492)
ballerina-platform/ballerina-lang: stdlib/messaging/nats/src/main/ballerina/nats/Module.md

apis to publish messages to each server.

#### NATS Server

Once connected, publishing is accomplished via one of three methods.

1. Publish with subject and message content
```ballerina
nats:Producer producer = new(connection);
var result = producer->publish(subject, "hello world");
```

2. Publish with subject, message content and a subject for the receiver to replyTo.
```ballerina
nats:Producer producer = new(connection);
var result = producer->publish(subject, "hello world", replyTo = "replyTo");
```

3. Publish as a request that expects a reply.
```ballerina
nats:Producer producer = new(connection);
var reqReply = producer->request(subject, "hello world", 5000);
```

#### Streaming Server

Once connected to streaming server, publishing message is accomplished using following method.
```ballerina
nats:StreamingProducer producer = new(connection);
var result = producer->publish(subject, "hello world");
if (result is error) {
   io:println("Error occurred while publishing the message.");
} else {
   io:println("GUID "+result+" received for the produced message.");
}
```

> NOTE: Publish api supports `byte[], boolean, string, int, float, decimal, xml, json, record {}` message types.

Reviewer: ```suggestion
NOTE: Publish api supports the `byte[], boolean, string, int, float, decimal, xml, json, record {}` message types.
```

### Listening for Incoming Messages

> The Ballerina NATS module provides following mechanisms to listen for messages. Like message publishing, listening 

Reviewer 2: ```suggestion
The Ballerina NATS module provides the following mechanisms to listen to messages. Similar to message publishing, listening to messages
```

> also handles differently in NATS server and Streaming server.

Reviewer 2: ```suggestion
is also handled differently in NATS and streaming servers.
```

Reviewer: ```suggestion
is also handled differently in NATS and Streaming servers.
```

> #### NATS Server

Reviewer 2: ```suggestion
#### Listening to messages in a NATS server
```

```ballerina
// Initializes the NATS listener.
listener nats:Listener subscription = new(connection);

// Binds the consumer to listen to the messages published to the 'demo' subject.
@nats:SubscriptionConfig {
    subject: "demo"
}
service demo on subscription {

    resource function onMessage(nats:Message msg, string data) {
        // Prints the incoming message in the console.
        io:println("Subject : " + msg.getSubject());
        io:println("Message content : " + data));
    }

    resource function onError(nats:Message msg, nats:NatsError err) {
        io:println(err);
    }

}
```

> #### Streaming Server

Reviewer 2: ```suggestion
#### Listening to messages in a Streaming server
```

```ballerina

> // Initializes the NATS listener.

Reviewer 2: ```suggestion
// Initializes the Streaming listener.
```

listener nats:StreamingListener subscription = new(connection);

// Binds the consumer to listen to the messages published to the 'demo' subject.
@nats:StreamingSubscriptionConfig {
    subject: "demo"
}
service demo on subscription {

    resource function onMessage(nats:StreamingMessage msg, string data) {
        // Prints the incoming message in the console.
        io:println("Subject : " + msg.getSubject());
        io:println("Message content : " + data));
    }

    resource function onError(nats:StreamingMessage msg, nats:NatsError err) {
        io:println(err);
    }

}
```

> ## Advanced Usage

Reviewer 2: ```suggestion
## Advanced use cases
```

Reviewer: This is not really a "use case"

> ### TLS

Reviewer 2: ```suggestion
### Using the TLS protocol
```

> The Ballerina NATS module allows the use of the tls:// protocol in its urls. This setting expects secureSocket to be 

Reviewer 2: ```suggestion
The Ballerina NATS module allows the use of the tls:// protocol in its URLs. This setting expects a secure socket to be 
```

> set in connection configuration like below.

Reviewer 2: ```suggestion
set in the connection configuration as shown below.
```

## Module overview

NATS is a simple, secure, and high-performance open source messaging system for cloud-native applications.
The 'ballerina/nats' module provides the capability to connect with NATS and Streaming servers and perform 
the following functionalities.

- Point to point communication (Queues)
- Pub/Sub (Topics)
- Request/Reply

## Basic Usage

### Setting up the connection

First step is setting up the connection with the NATS/Streaming server. The following ways can be used to connect to a
NATS/Streaming server.

1. Connect to a server using the URL
```ballerina
nats:Connection connection = new("nats://localhost:4222");
```

2. Connect to one or more servers with a custom configuration
```ballerina

> nats:Connection connection = new("nats://serverone:4222,nats://servertwo:4222",  config = config);

Reviewer 3: ```suggestion
nats:Connection connection = new("nats://serverone:4222, nats://servertwo:4222",  config = config);
```

Also apply the latest changes for optional param.

```

### Publishing messages

Publishing messages is handled differently in the NATS server and Streaming server. The 'ballerina/nats' module provides different 
APIs to publish messages to each server.

#### Publishing messages to the NATS server

Once connected, publishing is accomplished via one of the below three methods.

1. Publish with the subject and the message content.
```ballerina
nats:Producer producer = new(connection);

> var result = producer->publish(subject, "hello world");

Reviewer 3: Shall we use the return type like error? instead of var? WDYT?

```

2. Publish with subject, message content and a subject for the receiver to replyTo.
```ballerina
nats:Producer producer = new(connection);
var result = producer->publish(subject, "hello world", replyTo = "replyTo");
```

3. Publish as a request that expects a reply.
```ballerina
nats:Producer producer = new(connection);
var reqReply = producer->request(subject, "hello world", 5000);
```

#### Streaming Server

Once connected to a streaming server, publishing messages is accomplished using the following method.
```ballerina
nats:StreamingProducer producer = new(connection);
var result = producer->publish(subject, "hello world");
if (result is error) {
   io:println("Error occurred while publishing the message.");
} else {
   io:println("GUID "+result+" received for the produced message.");
}
```

NOTE: Publish api supports the `byte[], boolean, string, int, float, decimal, xml, json, record {}` message types.

### Listening to incoming messages

The Ballerina NATS module provides the following mechanisms to listen to messages. Similar to message publishing, listening to messages
is also handled differently in NATS and Streaming servers.

> #### NATS Server

Reviewer: ```suggestion
#### Listening to messages from a NATS server
```

> ```ballerina

Reviewer 3: Need import statement here.

// Initializes the NATS listener.
listener nats:Listener subscription = new(connection);

// Binds the consumer to listen to the messages published to the 'demo' subject.
@nats:SubscriptionConfig {
    subject: "demo"
}
service demo on subscription {

    resource function onMessage(nats:Message msg, string data) {
        // Prints the incoming message in the console.
        io:println("Subject : " + msg.getSubject());
        io:println("Message content : " + data));
    }

    resource function onError(nats:Message msg, nats:NatsError err) {
        io:println(err);
    }

}
```

> #### Streaming Server

Reviewer: ```suggestion
#### Listening to messages from a Streaming server
```

```ballerina

> // Initializes the NATS listener.

Reviewer: ```suggestion
// Initializes the NATS Streaming listener.
```
